Adding support for Typeinference based on runtime variable values

powercode opened the pull request:
fixes #2567 

Adding TypeInferenceVisitor (transforming GetInferredType method of corresponding types)
Removing Ast.GetInferredType hierarchy of virtual methods
Adding calls to SafeExpressEval when TypeInference is used in tabexpansion to allow
use of runtime variables when inferring type.

Commit:
Adding TypeInferenceVisitor with methods transformed from Ast.GetInferredType

Commit:
Removing the Ast.GetInferredType virtual methods from all ASTs

also delegating AddCommandWithPreference

Commit:
Replace CompletionExecutionHelper with PowerShellExecutionHelper

Commit:
Moving CurrentTypeDefinitionAst to TypeInferenceContext

Commit:
Enabling TypeInference use of runtime SafeExprEval for completion

Commit:
Adding tests and fixing issues

msftclas commented:
Hi __@powercode__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft and real humans are currently evaluating your PR.</p>

TTYL, MSBOT;

lzybkr commented on src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:
```diff
@@ -29,7 +29,7 @@
 {
     /// <summary>
-    /// 
+    ///
```
Please avoid formatting changes in a PR with meaningful code changes - it makes reviewing the code differences harder.

powercode commented:
@lzybkr I'm between a rock and a hard place here, having to explain the new push with either arrogance or incompetence :) It was the latter. I heard you the first time, but missed how I had Beyond compare configured to ignore whitespace.
Fixed now

vors commented:
@powercode can you, please, rebase and resolve the conflicts?

vors commented on src/System.Management.Automation/engine/CommandCompletion/CommandCompletion.cs:
```diff
@@ -643,7 +643,7 @@ private static List<CompletionResult> InvokeLegacyTabExpansion(PowerShell powers
             var lastword = LastWordFinder.FindLastWord(legacyInput, out replacementIndex, out quote);
             replacementLength = legacyInput.Length - replacementIndex;
-            var helper = new CompletionExecutionHelper(powershell);
+            var helper = new PowerShellExecutionHelper(powershell);
```
This replacement is reasonable from the commits history point of view, but in the final diff, I don't see why do we want to remove one class/file and replace it with another one with the similar duties. Maybe we should keep the same name, so the caller doesn't need to change?

It will also help, if PowerShell team will decide to bring some changes back to windows, for that see https://github.com/PowerShell/PowerShell/blob/master/docs/dev-process/map-json-files.md

vors commented on src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs:
```diff
@@ -96,8 +96,14 @@ private static bool IsCursorOutsideOfExtent(IScriptPosition cursor, IScriptExten
         }
 
-        internal CompletionContext CreateCompletionContext(ExecutionContext executionContext)
+        internal CompletionContext CreateCompletionContext(CompletionContext completionContext)
```
The signature suggests that it should be some sort of a copy-ctor, but it's not.

vors commented on src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:
```diff
@@ -0,0 +1,1413 @@
+        /// <summary>
+        /// No runtime use is allowed
+        /// </summary>
+        None,
```
Should it be explicit `0` for the value?

powercode replied:
fixing

vors commented on src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:
```diff
@@ -0,0 +1,1413 @@
+            return false;
+        }
+
+        public IList<object> GetMembersByInferredType(PSTypeName typename, bool @static, Func<object, bool> filter)
```
Please, avoid using `@static`, `@class`, etc for variable names. Replace by `isStatic`.

vors replied:
`GetMembersByInferredType` method is too big. It takes 3 monitors on my desktop. Please, break it down into sub-routines: i.e. every if-else body can be a sub-routine.

powercode replied:
I just followed the style in ast.cs. Fixing.

vors commented on src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:
```diff
@@ -0,0 +1,1413 @@
+                }
+                else
+                {
+                    elementTypes = typename.Type.GetInterfaces().Where(
```
Code in this file may be executed on every keystroke in some scenarios, so it's better avoid LINQ and extensive lambda creation.

https://github.com/PowerShell/PowerShell/blob/master/docs/dev-process/coding-guidelines.md#performance-considerations

vors commented on src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:
```diff
@@ -0,0 +1,1413 @@
+            return results;
+        }
+
+        internal static PowerShell AddCommandWithPreferenceSetting(PowerShell powershell, string command, Type type = null)
```
This method copy-pasted from `src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs`, why?

powercode replied:
Forgot I had done that. I was thinking about dependencies. I think it's ok for Completion to have a dependency on TypeInference, but not the other way. So I think I had the idea to move it to TypeInference, and call it from Completion, but never got around to it.

vors commented on src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:
```diff
@@ -0,0 +1,1413 @@
+            return powershell;
+        }
+
+        private static bool ParseCimCommandsTypeName(PSTypeName typename, out string cimNamespace, out string className)
```
Currently there is no CIM on Unix. Should we add conditional compilation?

vors commented on src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:
```diff
@@ -0,0 +1,1413 @@
+            return TypeInferenceContext.EmptyPSTypeNameArray;
+        }
+
+        private IEnumerable<PSTypeName> InferTypesFrom(CommandAst commandAst)
```
This method is too big, please break it down

vors commented on src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:
```diff
@@ -0,0 +1,1413 @@
+            }
+        }
+
+        private IEnumerable<PSTypeName> InferTypesFrom(MemberExpressionAst memberExpressionAst)
```
This method is too big

vors commented on src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:
```diff
@@ -0,0 +1,1413 @@
+            }
+        }
+
+        private IEnumerable<PSTypeName> InferTypeFrom(VariableExpressionAst variableExpressionAst)
```
This method is too big

powercode commented on src/System.Management.Automation/utils/PowerShellExecutionHelper.cs:
```diff
@@ -1,34 +1,24 @@
-
-/********************************************************************++
-Copyright (c) Microsoft Corporation.  All rights reserved.
---********************************************************************/
```
Because it is no longer used for completion, but for type inference.
Jason and I had a discussion, where he pointed out that for example Script Analyzer would be interested in type inference.

Previously, there was an unfortunate dependency in the AST on Completion, and this was in an effort to make it clear that the class was not only used by completion.

vors replied:
Ok, lets rename it.

* I think `PowerShellExecutionHelper.cs` name falls into another extreme: the name is too generic and doesn't really describe what's inside the file, but naming is hard and it's a nit-pick.
* Why remove the copyright?

powercode replied:
Copyright is just my mistake. Restoring.

powercode commented on src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:
```diff
@@ -0,0 +1,1413 @@
+    /// Enum describing permissions to use runtime evaluation during type inference
+    /// </summary>
+    [Flags]
+    public enum TypeInferenceRuntimePermissions {
```
I think I would like you to have this discussion with @lzybkr. I have to admit that I don't have a clear view regarding what should be made public and what shouldn't. 
It is clearly easier to make internal api:s public later on, then to go the other way.

vors replied:
@lzybkr ? 

Following other visitors (i.e.`SemanticChecks`, ) I think we should start with `internal`.

lzybkr replied:
The goal here is to end up with a public type inference API.  Script analyzer is one obvious tool that would benefit (in fact they have their own implementation, and a shared implementation would be better I think.)

powercode commented on src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:
```diff
@@ -0,0 +1,1413 @@
+            return (IEnumerable<PSTypeName>)res;
+        }
+
+        public bool TryGetRepresentativeTypeNameFromExpressionSafeEval(ExpressionAst expression, out PSTypeName typeName)
```
fixing

vors commented:
Waiting on addressing the code-review comments

powercode commented on src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs:
```diff
@@ -1167,7 +1177,7 @@ private List<CompletionResult> GetResultForString(CompletionContext completionCo
                 var analysis = new CompletionAnalysis(_ast, _tokens, _cursorPosition, _options);
-                var subContext = analysis.CreateCompletionContext(completionContext.ExecutionContext);
+                var subContext = analysis.CreateCompletionContext(completionContext);
                 subContext.Helper = completionContext.Helper;
```
indeed we should :)

vors commented on src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:
```diff
@@ -6463,26 +6464,7 @@ private static List<CompletionResult> GetSpecialHashTableKeyMembers(params strin
 
         internal static PowerShell AddCommandWithPreferenceSetting(PowerShell powershell, string command, Type type = null)
```
This is internal method. Should we just remove it completely?

vors commented on src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:
```diff
@@ -0,0 +1,1505 @@
+            return dynamicKeywordAst.CommandElements[0].Accept(this);
+        }
+    }
+}
```
New line at the end of file to make git happy

lzybkr replied:
Now that we have code coverage data, it would be great to get some great coverage on this new file.

powercode replied:
fixing

lzybkr commented on src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:
```diff
@@ -0,0 +1,1505 @@
+        public bool TryGetRepresentativeTypeNameFromExpressionSafeEval(ExpressionAst expression, out PSTypeName typeName)
+        {
+            typeName = null;
+            if (!RuntimePermissions.HasFlag(TypeInferenceRuntimePermissions.AllowSafeEval))
```
We should avoid Enum.HasFlag, it's much, much more expensive than a simple bitwise test.

See the implementation here: https://github.com/dotnet/coreclr/blob/e67851210d1c03d730a3bc97a87e8a6713bbf772/src/vm/reflectioninvocation.cpp#L3699

powercode replied:
fixing

lzybkr commented on src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:
```diff
@@ -0,0 +1,1505 @@
+                var pathValue = ((StringConstantExpressionAst)pathArgumentPair.Argument).Value;
+                try
+                {
+                    commandInfo = commandInfo.CreateGetCommandCopy(new object[] { "-" + pathParameterName, pathValue });
```
Some comments on what this is all about would be nice, right?  (It's probably my fault, so hopefully I remember correctly).

The `OutputType` on cmdlets like `Get-ChildItem` may depend on the path. The `CmdletInfo` returned based on just the command name will specify returning all possibilities, e.g. certificates, environment, registry, etc.

If you specified `-Path`, the list of `OutputType` can be refined, but we have to make a copy of the `CmdletInfo` object this way to get that refinement.

powercode replied:
fixing

powercode commented:
If I understand correctly, we still have to 

- determine a good name for PowerShellExecutionHelper
- decide if the public API is good enough or what changes should be made to it.
- anything else?

vors commented:
The iterations on this PR are very unclear. It's impossible to tell what changed from the previos time, because history is constantly rewritten.
I think @daxian-dbw already proposed to encourage keep the history as-is in the PR and only rewrite it when the review is done. I think it would help a lot to follow such proposition in this particular case.
I would not have time in the next 2 week to do the review, so clearing the assignment.

Commit:
adding resharper settings to gitignore

Commit:
working around platform path diffs

Commit:
Skipping CIM test on non-windows

powercode commented:
@vors  Sorry about all the rebasing. We had a discussion offline about this, that wasn´t conclusive. But in this case, with so many iteration, it gets really bad. I think I agree with @daxian-dbw. Even though it adds burden when doing the final merge, the ease of reviewing makes up for it.

lzybkr commented on src/System.Management.Automation/engine/parser/ast.cs:
```diff
@@ -9580,7 +8715,7 @@ internal bool IsSafeVariableReference(HashSet<string> validVariables, ref bool u
 
-        private bool AstAssignsToSameVariable(Ast ast)
+        internal bool AstAssignsToSameVariable(Ast ast)
         {
```
The only reference to this was in `GetInferredType` - so maybe move this method too.

powercode replied:
moved

daxian-dbw replied:
This method is moved to `TypeInferenceVisitor.cs`, so this method should be removed from `ast.cs`.

lzybkr commented on src/System.Management.Automation/engine/CommandCompletion/PseudoParameterBinder.cs:
```diff
@@ -1338,79 +1335,76 @@ private bool PrepareCommandElements(ExecutionContext context)
                 var converterType = Type.GetType(Utils.WorkflowType);
-                if (converterType != null)
+                var activityParameters = (Dictionary<string, Type>) converterType?.GetMethod("GetActivityParameters").Invoke(null, new object[] { _commandAst });
+                if (activityParameters != null)
```
In the future, I'd prefer to see language usage changes like this as a single PR, and I'd prefer to see 1 commit per language usage change.

daxian-dbw replied:
The diffs look horrible in Github, but is not that bad in codeflow.

lzybkr commented on src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:
```diff
@@ -6473,26 +6474,7 @@ private static List<CompletionResult> GetSpecialHashTableKeyMembers(params strin
-                .AddParameter("Debug", false);
-
-            return powershell;
+            return TypeInferenceContext.AddCommandWithPreferenceSetting(powershell, command, type);
```
Moving this method to `TypeInferenceContext` feels wrong - this is a generic utility method for completion, it's used for more than type inference.

powercode replied:
Agree. What is a good place for such generic utilities? 

And should it be an extension method on PowerShell for discoverability?

lzybkr commented on src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:
```diff
@@ -0,0 +1,1531 @@
+    }
+
+    /// <summary>
+    /// static class containting methods to work with type inference of abstract syntax trees
```
>containting [](start = 21, length = 11)

containing

lzybkr commented on src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:
```diff
@@ -0,0 +1,1531 @@
+        /// Infers the type that the result of executing a statement would have
+        /// </summary>
+        /// <param name="ast">the ast to infer the type from</param>
+        /// <param name="evalPersmissions">The runtime usage permissions allowed during type inference</param>
```
>evalPersmissions [](start = 25, length = 16)

evalPermissions

lzybkr commented on src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:
```diff
@@ -0,0 +1,1531 @@
+    /// static class containting methods to work with type inference of abstract syntax trees
+    /// </summary>
+    public static class AstTypeInference
+    {
```
I'd like to keep the api internal for now.
In the short term, we are considering adding an "experimental" option - and when we have that, then I'd like to make this public but experimental, meaning external code needs to opt-in explicitly or the api wouldn't work.

powercode replied:
Do we have a good way to test internal APIs? All the test I've written are based on AstTypeInference.InferType.

lzybkr replied:
Not really - the simplest is to make the type internal, but everything else public - then you use a little bit of reflection to get the type, and after that, everything just works like normal.

lzybkr commented on src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:
```diff
@@ -0,0 +1,1531 @@
+                if (propertyMember != null)
+                {
+                    if (propertyMember.IsStatic == isStatic)
+                    {
```
This can be simplified to `add = propertymember.IsStatic == isStatic`, with a similar change in the else branch, then the default initialization isn't necessary anymore.

lzybkr commented on src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:
```diff
@@ -0,0 +1,1531 @@
+            {
+                if (filterToCall == null)
+                    filterToCall = o => !IsMemberHidden(o);
+                else
```
Braces around the if/else body are missing, or maybe convert this to a ternary expression if that's cleaner.

lzybkr commented on src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:
```diff
@@ -0,0 +1,1531 @@
+        private readonly TypeInferenceContext _context;
+
+
+        public TypeInferenceVisitor(TypeInferenceContext context)
```
Delete extra blank line.

lzybkr commented on src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:
```diff
@@ -0,0 +1,1531 @@
+        object ICustomAstVisitor.VisitExpandableStringExpression(ExpandableStringExpressionAst expandableStringExpressionAst)
+        {
+            return new[] { new PSTypeName(typeof(string)) };
+        }
```
When the type is a constant like this, we should probably return a static to minimize garbage.

powercode commented:
I'm addressing the comments, but need advice on how to proceed with the tests that are using the public API.

powercode commented:
@lzybkr Can we have a test assembly with InternalsVisibleTo SMA which just delegate to the internal API?

Commit:
AstTypeInference -> internal + adressing review comments

Using reflection in tests to call the internal methods.

powercode commented:
@lzybkr  Review comments addressed.

Commit:
Moving AddCommandWithPreferenceSetting

New location is in PowerShellExecutionHelper.cs

powercode commented:
Tests are now using reflection. I think all comments are addressed.

daxian-dbw commented on src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs:
```diff
@@ -1167,8 +1172,7 @@ private List<CompletionResult> GetResultForString(CompletionContext completionCo
                 var analysis = new CompletionAnalysis(_ast, _tokens, _cursorPosition, _options);
-                var subContext = analysis.CreateCompletionContext(completionContext.ExecutionContext);
-                subContext.Helper = completionContext.Helper;
+                var subContext = analysis.CreateCompletionContext(completionContext.TypeInferenceContext.Helper.CurrentPowerShell);
```
> completionContext.TypeInferenceContext.Helper.CurrentPowerShell

Why not just `completionContext.Helper.CurrentPowerShell`?

On the second thought, a single `Helper` was shared by `completionContext` and `subContext` previously, now new objects `TypeInferenceContext` and `PowerShellExecutionHelper` will be created. It looks to me `TypeInferenceContext` is stateless and can be shared by two `CompletionContext`, so it seems reasonable to continue shared the same objects.

How about add an overload of `CompletionContext.CreateCompletionContext`? Like this:
```CSharp
internal CompletionContext CreateCompletionContext(PowerShell powerShell)
{
    var typeInferenceContext = new TypeInferenceContext(powerShell)
    {
        CurrentTypeDefinitionAst = Ast.GetAncestorTypeDefinitionAst(asts.Last())
    };
    InitCompletionContext(typeInferenceContext)
}

internal CompletionContext CreateCompletionContext(TypeInferenceContext typeinferenceContext)
{
    InitCompletionContext(typeInferenceContext)
}

internal void InitCompletionContext(TypeInferenceContext typeinferenceContext)
{
    // Existing init code
    ...
    return new CompletionContext
    {
        ...
        ExecutionContext = typeinferenceContext.ExecutionContext,
        TypeInferenceContext = typeInferenceContext,
        Helper = typeInferenceContext.Helper,
        ...
    }
}
```
So `subContext` can be created via `var subContext = analysis.CreateCompletionContext(completionContext.TypeInferenceContext)` and share the same `TypeInferenceContext` and `PowerShellExecutionHelper` with `completionContext`.

powercode replied:
Makes sense. Fixing

daxian-dbw commented on src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs:
```diff
@@ -590,7 +595,7 @@ internal List<CompletionResult> GetResultHelper(CompletionContext completionCont
                     //   new-object System.Drawing.Point -prop @{ X=  -> Tab should not complete
                     // Note: This check works when all statements preceding the last are complete,
-                    //       but if a preceding statement is incomplete this test fails because
+                    //       but if a preceding statement is incomplete this test fails because 
```
Can you please remove the extra space at the end?

daxian-dbw commented on src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs:
```diff
@@ -1263,7 +1267,7 @@ private List<CompletionResult> GetResultForString(CompletionContext completionCo
         /// Results are generated based on DynamicKeywords matches given identifier.
         /// For example, following "Fi" matches "File", and "Us" matches "User"
-        ///
+        /// 
```
Please remove the extra space.

daxian-dbw commented on src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs:
```diff
@@ -1272,7 +1276,7 @@ private List<CompletionResult> GetResultForString(CompletionContext completionCo
         ///         }
         ///     }
-        ///
+        /// 
```
Extra space here.

daxian-dbw commented on src/System.Management.Automation/engine/CommandCompletion/PseudoParameterBinder.cs:
```diff
@@ -1451,32 +1445,36 @@ private CommandProcessorBase PrepareFromAst(ExecutionContext context, out string
-            int resolvedAliasCount = 0;
-            while (exportVisitor.DiscoveredAliases.TryGetValue(resolvedCommandName, out alias))
+            resolvedCommandName = _commandAst.GetCommandName();
+            if (resolvedCommandName != null)
```
Good catch 👍

Commit:
Sharing helper objects instead of creating new

Addressing review comments from @daxian-dbw
Removing orphaned AstAssignsToSameVariable method

powercode commented:
@daxian-dbw Good suggestions! Have a look and see if I've understood you correctly.

daxian-dbw commented on src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:
```diff
@@ -5377,9 +5377,10 @@ internal static IEnumerable<object> GetMembersByInferredType(PSTypeName typename
             if (typename.Type != null)
             {
-                if (context.CurrentTypeDefinitionAst == null || context.CurrentTypeDefinitionAst.Type != typename.Type)
+                if (typeInferenceContext.CurrentTypeDefinitionAst == null || typeInferenceContext.CurrentTypeDefinitionAst.Type != typename.Type)
```
It looks a lot code is duplicated between here and TypeInferenceContext.cs. Duplicate code should be removed.

daxian-dbw commented on src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:
```diff
@@ -0,0 +1,1572 @@
+                var baseTypeName = baseType.TypeName as TypeName;
+                if (baseTypeName == null) continue;
+                var baseTypeDefinitionAst = baseTypeName._typeDefinitionAst;
+                results.AddRange(GetMembersByInferredType(new PSTypeName(baseTypeDefinitionAst), isStatic, filterToCall));
```
`class foo : Hashtable {}`
In this case, `baseType._typeDefinitionAst` is null, and thus `new PSTypeName(baseTypeDefinitionAst)` will throw.

powercode commented on src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:
```diff
@@ -0,0 +1,1572 @@
+    /// <summary>
+    /// Enum describing permissions to use runtime evaluation during type inference
+    /// </summary>
+    [Flags]
```
No, it should not

powercode commented on src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:
```diff
@@ -0,0 +1,1572 @@
+        /// <returns></returns>
+        public static IList<PSTypeName> InferTypeOf(Ast ast, PowerShell powerShell)
+        {
+            return InferTypeOf(ast, TypeInferenceRuntimePermissions.None);
```
check!

daxian-dbw commented on src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:
```diff
@@ -0,0 +1,1572 @@
+        internal IEnumerable<PSTypeName> InferType(Ast ast, TypeInferenceVisitor visitor)
+        {
+            var res = ast.Accept(visitor);
+            if (res == null)
```
Just curious, in what case will `res` be null?

powercode replied:
I don't think it ever will. This is a case where I wish I could express non-nullable reference types in the type system, but since the signature is an object, I check it. Maybe replace with an assert?

daxian-dbw replied:
> I don't think it ever will.

That's my feeling by looking at the code :) If that's the case, I prefer to an assert.

lzybkr replied:
Where possible, I try to return some sentinel like an empty static array instead of null - when you do that consistently, you can end up with simpler code that doesn't need to test for `null`.

powercode commented on src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:
```diff
@@ -0,0 +1,1572 @@
+        /// <param name="powerShell">the instance of powershell to user for expression evalutaion needed for type inference</param>
+        /// <param name="evalPersmissions">The runtime usage permissions allowed during type inference</param>
+        /// <returns></returns>
+        public static IList<PSTypeName> InferTypeOf(Ast ast, PowerShell powerShell,TypeInferenceRuntimePermissions evalPersmissions)
```
Changing to` _powerShell.Runspace?.ExecutionContext;`
Callers will have to check.

daxian-dbw replied:
Better add comments and in the constructors to call it out.
Do you expect the passed in powershell instance to associate with a remote runspace or a runspace pool? For the usage of it in our tab completion code, it will always be created from `powershell.Create(CurrentRunspace)`, but I'm having trouble to figure out the scenarios it will be used outside tab completion code.

powercode replied:
@lzybkr has hopes that it will be used by Script Analyzer too.

powercode commented on src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:
```diff
@@ -0,0 +1,1572 @@
+            }
+        }
+
+        struct ContextPermissionScope : IDisposable
```
Restructuring the code instead and removing it. Good point!

daxian-dbw commented on src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:
```diff
@@ -0,0 +1,1572 @@
+        public bool TryGetRepresentativeTypeNameFromExpressionSafeEval(ExpressionAst expression, out PSTypeName typeName)
+        {
+            typeName = null;
+            if ((RuntimePermissions & TypeInferenceRuntimePermissions.AllowSafeEval) != TypeInferenceRuntimePermissions.AllowSafeEval)
```
If the `Flag` attribute is not necessary, then it's more readable to change this line to
```
RuntimePermissions != TypeInferenceRuntimePermissions.AllowSafeEval
```

daxian-dbw commented:
@lzybkr LINQ is used in many places in the changes, some are new and some are from existing code. 
Tab completion code is perf-sensitive, so should we always disallow LINQ usages? What would be the guideline?

Commit:
powershell -> use context.Helper with AddCommandWithPreferenceSetting

Adding an overload on AddCommandWithPreferenceSetting for
PowerShellExecutionHelper. Updating usages.

powercode commented:
I discussed this with @lzybkr and @vors  on gitter. I wanted to use a preallocated list that was filled in by the visitor. Mostly for ease of debugging, but also for performance. My impression was that Sergei liked the readablilty of linq and Jason was slightly in favor of using the list.

I'm open to both. It would be nice to see some performance data on the existing code to see if it is actually needed. The data sets are seldom large here. 
Or maybe they are when analyzing scripts?

powercode commented:
By the way, @daxian-dbw : what a great review! Working through the issues.

Commit:
Removing trailing whitespace

lzybkr commented:
When LINQ gets used inappropriately - you end up with many allocations, and that affects GC at some point, possibly during your scenario, possibly not.

This is unrelated to `yield` vs. a list.

Commit:
Adding -Force to Update-TypeData

powercode commented:
I have chosen to not make to many changes since the main job of this PR was to enable a new feature.
I tried to keep the code fairly similar to the original to make it easier to see my changes but that is soon a moot point :)
Maybe things like Linq usage can be a separate PR?

powercode commented on src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:
```diff
@@ -0,0 +1,1572 @@
+                foreach (var cimClass in classes.Select(PSObject.Base).OfType<CimClass>())
+                {
+                    results.AddRange(filterToCall != null
+                        ? cimClass.CimClassProperties.Where(filterToCall)
```
unrolling

powercode commented on src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:
```diff
@@ -0,0 +1,1572 @@
+                }
+            }
+            IEnumerable<Type> elementTypes;
+            if (typename.Type.IsArray)
```
Maybe it is the naming that is bad. The whole scenario seems to be to get the type from enumerables and arrays

Commit:
Addressing review comments

Commit:
Removing code duplication

Commit:
GetMembersByInferredType -> TypeInferenceContext instance

powercode commented:
I believe I have addressed all comments except for LINQ usage.

daxian-dbw commented:
The impact of LINQ usage doesn't show up immediately. Initially, the perf difference won't be measurable, but as more LINQ usages put in the code gradually, the perf diff will show up. By that time, it would be difficult to refactor the code to replace those uses.
So I vote for minimizing LINQ uses in perf-sensitive codes like Parser, Compiler and Auto-Completion.

> Maybe things like Linq usage can be a separate PR?

I agree to make the LINQ usage changes a separate PR.

lzybkr commented on test/powershell/engine/Api/TypeInference.Tests.ps1:
```diff
@@ -0,0 +1,859 @@
+
+    }
+
+    It "Inferes type from integer" {
```
>Inferes [](start = 8, length = 7)

`Infers` - I could overlook this once, but it's copied many times in this file.

lzybkr commented on test/powershell/engine/Api/TypeInference.Tests.ps1:
```diff
@@ -0,0 +1,859 @@
+    It "Inferes type from array IndexExpresssion" {
+        $res = [AstTypeInference]::InferTypeOf( { (1, 2, 3)[0] }.Ast)
+        $res.Count | Should Be 1
+        $res.Name | Should be 'System.object'
```
>'System.object' [](start = 30, length = 15)

Hopefully there is a todo somewhere so this result comes out as `System.Int32`.

powercode replied:
Do you expect me to fix this in PR?
I don't have a todo for this. I just tried to capture the current general behavior of arrays in a test.

powercode replied:
Then there are other cases too:
```
(<expr>)[0]
(<expr>, <expr2>)[0..1]
(<expr>, <expr2>)[0,"1"]
```
Do we have any nice existing facilities to get the values of an index's target expression?
I guess we want to convert it to an int[], and use the indexes to get the elements of the target (in case it is an arrayliteral)? And return the types inferred from those of the elements that are in range?

lzybkr replied:
It would be nice if you could at least open an issue to track the areas you know we're not currently wrong, but can somewhat easily do better.

Commit:
Correct bad spelling of 'Infers' in tests

powercode commented:
Opened #3857 to track improvment of inference for IndexExpresssions.

powercode commented:
Opened #3858 to track the removal of LINQ from TypeInferenceVisitor. @daxian-dbw

vors commented:
> My impression was that Sergei liked the readablilty of linq and Jason was slightly in favor of using the list.

Not really: I was in favor of stateless passing everything around as parameters, linq was not discussed.

iSazonov commented:
@powercode Thanks for the great work!
